src/experimental/patterns/ec2-app.ts

```diff
@@ -0,0 +1,309 @@
+  private readonly asgToParamMap: Map<string, CfnParameter>;
+  private static instance: HorizontallyScalingDeploymentProperties | undefined;
+
+  private constructor(scope: GuStack) {
+    this.stack = scope;
+    this.asgToParamMap = new Map();
+  }
+
+  public static getInstance(stack: GuStack): HorizontallyScalingDeploymentProperties {
+    if (!this.instance || !isSingletonPresentInStack(stack, this.instance)) {
+      this.instance = new HorizontallyScalingDeploymentProperties(stack);
+    }
+
+    return this.instance;
+  }
+
+  public visit(construct: IConstruct) {
+    if (construct instanceof CfnScalingPolicy) {
+      const { node } = construct;
+      const { scopes, path } = node;
+
+      /*
+      Requiring a `CfnScalingPolicy` to be created in the scope of a `GuAutoScalingGroup`
+      is the most reliable way to associate the two together.
+
+      Though the `autoScalingGroupName` property is passed to a `CfnScalingPolicy` when instantiating,
+      this does not create a concrete link as AWS CDK sets the value to be the ASG's `Ref`.
+      That is, it is a `Token` until it is synthesised.
+      This is even if the ASG has an explicit name set.
+
+      See also:
+        - https://docs.aws.amazon.com/cdk/v2/guide/tokens.html
+        - https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-autoscaling-autoscalinggroup.html#aws-resource-autoscaling-autoscalinggroup-return-values
+        - https://github.com/aws/aws-cdk/blob/f6b649d47f8bc30ca741fbb7a4852d51e8275002/packages/aws-cdk-lib/aws-autoscaling/lib/auto-scaling-group.ts#L1560
+       */
+      const autoScalingGroup = scopes.find((_): _ is GuAutoScalingGroup => _ instanceof GuAutoScalingGroup);
+
+      if (!autoScalingGroup) {
+        throw new Error(
+          `Failed to detect the autoscaling group relating to the scaling policy on path ${path}. Was it created in the scope of a GuAutoScalingGroup?`,
+        );
+      }
+
+      const cfnAutoScalingGroup = autoScalingGroup.node.defaultChild as CfnAutoScalingGroup;
+      const currentRollingUpdate = cfnAutoScalingGroup.cfnOptions.updatePolicy?.autoScalingRollingUpdate;
+
+      if (currentRollingUpdate) {
+        /*
+        An autoscaling group that horizontally scales should not explicitly set `Desired`,
+        as a rolling update will set the current `Desired` back to the template version,
+        undoing any changes that a scale-out event may have done.
+         */
+        cfnAutoScalingGroup.desiredCapacity = undefined;
+
+        const asgNodeId = autoScalingGroup.node.id;
+
+        if (!this.asgToParamMap.has(asgNodeId)) {
+          this.asgToParamMap.set(
+            asgNodeId,
+            new CfnParameter(this.stack, `MinInstancesInServiceFor${autoScalingGroup.app}`, {
+              type: "Number",
+              default: parseInt(cfnAutoScalingGroup.minSize),
+              maxValue: parseInt(cfnAutoScalingGroup.maxSize) - 1,
+            }),
```

Comment:
If the `app` contains hyphens (`-`), or other illegal characters for a CFN Parameter, they'll get stripped out. This means Riff-Raff cannot reliably just find CFN Parameters that start with `MinInstancesInServiceFor`. Therefore Riff-Raff's logic would be something like this:
1. Find all `AWS::AutoScaling::ScalingPolicy`s in the CloudFormation template that has `MinInstancesInService` set via a CloudFormation Parameter
2. Find current `DesiredCapacity` of ASG that relates to the Scaling Policy
3. Set the CFN Parameter identified in step 1 to the value identified in step 2

Reply:
I wonder if our `Aspect` could add [`Outputs`](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/outputs-section-structure.html) for Riff-Raff to read. 

This would mean that Riff-Raff does not need to parse the full template or lookup the associated ASG name (which will not be included in the template unless [this property](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-autoscaling-autoscalinggroup.html#cfn-autoscaling-autoscalinggroup-autoscalinggroupname) is set).

For example:

```
Outputs:
  MinInstancesInServiceForAppA:
    Description: The AutoScaling group name associated with the MinInstancesInServiceForAppA parameter.    
    Value: !Ref AsgForAppA 
  MinInstancesInServiceForAppB:
    Description: The AutoScaling group name associated with the MinInstancesInServiceForAppB parameter.    
    Value: !Ref AsgForAppB
```

I think we can obtain these outputs pretty easily via the existing [`describeStack` function](https://github.com/guardian/riff-raff/blob/d02c0e5835a20d8389a97a19df896ebc4bed2f15/magenta-lib/src/main/scala/magenta/tasks/AWS.scala#L854-L868) and IIUC [using `!Ref` here](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-autoscaling-autoscalinggroup.html#aws-resource-autoscaling-autoscalinggroup-return-values) will mean that we can avoid looking up the ASG name ourselves.

Reply:
Oh interesting idea! Tempted to make these changes in a separate PR once we've know more about what the changes in Riff-Raff would look like. WDYT?